hrabsence/CRM/HRAbsence/Form/AbsenceRequest.php

```diff
@@ -517,320 +541,410 @@ static function formRule($fields, $files, $self) {
+        }
+      }
+
+      $this->updateCustomFields($result);
+
+      if ($this->_sendTemplateParams['sendMail']) {
+        self::sendAbsenceMail($this->_mailprm, $this->_sendTemplateParams);
+      }
+    } else {
+      if (CRM_Utils_Request::retrieve('aid', 'Positive', $this)) {
+        $activityIDs = CRM_Utils_Request::retrieve('aid', 'Positive', $this);
+      }
+
+      if (array_key_exists('_qf_AbsenceRequest_done_cancelabsence', $this->_submitValues)) {
+        $this->updateActivitiesStatus('Cancelled');
+        $statusMsg = ts('Absence(s) have been Cancelled');
+      } elseif (array_key_exists('_qf_AbsenceRequest_done_cancel', $this->_submitValues)) {
+        $session->pushUserContext(CRM_Utils_System::url('civicrm/absences', "reset=1&cid={$this->_targetContactID}", false, 'hrabsence/list'));
+      }
+      CRM_Core_Session::setStatus($statusMsg, '', 'success');
+      $session->pushUserContext(CRM_Utils_System::url('civicrm/absence/set', "reset=1&action=view&aid={$activityIDs}"));
     }
-    $absentDateDurations = array();
-    $activityStatus = CRM_HRAbsence_BAO_HRAbsenceType::getActivityStatus('name');
-    $activityStatusId['status_id'] = CRM_Utils_Array::key('Completed', $activityStatus);
+  }
 
-    if (!empty($submitValues['date_values'])) {
-      foreach (explode('|', $submitValues['date_values']) as $key => $dateString) {
+  /**
+   * Builds array of template's parameters to be passed to sendAbsenceMail function
+   * 
+   * @return array
+   *   Template parameters to build e-mail message to be sent to contacts
+   */
+  private function buildSendTemplateParams() {
+    $activityStatus = empty($this->_activityStatuses) ? 
+      CRM_HRAbsence_BAO_HRAbsenceType::getActivityStatus('name') : 
+      $this->_activityStatuses;
+
+    $absentDateDurations = [];
+
+    // Build array of absences per day
+    if (!empty($this->_submitValues['date_values'])) {
+      foreach (explode('|', $this->_submitValues['date_values']) as $dateString) {
         if ($dateString) {
           $values = explode('(', $dateString);
           $date = CRM_Utils_Date::processDate($values[0]);
           $valuesDate = explode(':', $dateString);
           $absentDateDurations[$date]['duration'] = (int) $valuesDate[1];
-          if ((isset($valuesDate[2]) && $valuesDate[2] == 1 && $this->_showhide == 1 && array_key_exists('_qf_AbsenceRequest_done_save', $submitValues)) || (isset($valuesDate[2]) && $valuesDate[2] == 0 && $this->_showhide == 0 && array_key_exists('_qf_AbsenceRequest_done_save', $submitValues))) {
+          if ((isset($valuesDate[2]) && $valuesDate[2] == 1 && $this->_showhide == 1 && array_key_exists('_qf_AbsenceRequest_done_save', $this->_submitValues)) || (isset($valuesDate[2]) && $valuesDate[2] == 0 && $this->_showhide == 0 && array_key_exists('_qf_AbsenceRequest_done_save', $this->_submitValues))) {
```

Comment:
this part is so crowded ... please put all **elseifs** on a new line and leave space before the **if**